File: 8. One More Time/finitefield/Java/polynomial.java
package Java;
import java.io.File;
import java.io.FileNotFoundException;
import java.util.ArrayList;
import java.util.List;
import java.util.Scanner;

public class polynomial {

    public static int mod(int a, int p) {
        return (a % p + p) % p;
    }

    public static int add(int a, int b, int p) {
        return mod(a+b,p);
    }

    public static int sub(int a, int b, int p) {
        return mod(a-b,p);
    }

    public static int mul(int a, int b, int p) {
        return mod(a*b,p);
    }

    public static int inv(int a, int p) {

        int t = 0, newt = 1;
        int r = p, newr = a;

        while (newr != 0) {
            int quotient = r / newr;
            int temp = t;
            t = newt;
            newt = temp - quotient * newt;
            temp = r;
            r = newr;
            newr = temp - quotient * newr;
        }

        if (t < 0) {
            t += p;
        }

        return t;

    }

    public static int div(int a, int b, int p) {
        return mul(a,inv(b,p),p);
    }

    public static int[] interpolate(int P, int d, List<Integer> x, List<Integer> y) {
        int[] p = new int[d];

        for (int i = 0; i < d; i++) {
            int product = 1;
            int[] t = new int[d];

            for (int j = 0; j < d; j++) {
                if (i != j) {
                    product = mul(product,sub(x.get(i), x.get(j),P),P);
                }
            }

            product = div(y.get(i), product,P);
            t[0] = product;

            for (int j = 0; j < d; j++) {
                if (i != j) {
                    for (int k = d - 1; k > 0; k--) {
                        t[k] = add(t[k],t[k - 1],P);
                        t[k - 1] = mul(t[k - 1],-x.get(j),P);
                    }
                }
            }

            for (int j = 0; j < d; j++) {
                p[j] = add(p[j],t[j],P);
            }
        }

        return p;
    }

    public static void main(String[] args) {
        if (args.length < 2) {
            System.out.println("Error: Please provide a prime number and a file path.");
            return;
        }
        
        Integer P = Integer.parseInt(args[0]);
        String path = args[1];
        File file = new File(path);
        if (!file.exists()) {
            System.out.printf("Error: The file at '%s' doesn't exist!\n", path);
            return;
        }

        List<Integer> x = new ArrayList<>();
        List<Integer> y = new ArrayList<>();
        
        try (Scanner scanner = new Scanner(file)) {
            while (scanner.hasNextLine()) {
                String line = scanner.nextLine();
                String[] split = line.split(" ");

                if (split.length != 2) {
                    System.out.println("Error: Invalid input format");
                    return;
                }

                x.add(Integer.parseInt(split[0]));
                y.add(Integer.parseInt(split[1]));
            }

            int[] p = interpolate(P,x.size(), x, y);
            for (int val : p) {
                System.out.print(val + " ");
            }
            System.out.println();
        } catch (FileNotFoundException e) {
            System.out.printf("Error: The file at '%s' doesn't exist!\n", path);
        } catch (NumberFormatException e) {
            System.out.println("Error: Invalid number format in the file");
        }
    }
}

File: 8. One More Time/finitefield/Ruby/polynomial.rb
# Galois field arithmetic operations
def gf_add(a, b, p)
  (a + b) % p
end

def gf_sub(a, b, p)
  (a - b + p) % p
end

def gf_mul(a, b, p)
  (a * b) % p
end

def gf_inv(a, p)
  raise 'Inverse does not exist' if a == 0
  (1..p-1).find { |i| (i * a) % p == 1 } || raise('Inverse not found')
end

def gf_div(a, b, p)
  gf_mul(a, gf_inv(b, p), p)
end

def interpolate(d, x, y, p)
  coeff = Array.new(d, 0)

  for i in 0...d
    product = 1
    t = Array.new(d, 0)

    for j in 0...d
      next if i == j
      product = gf_mul(product, gf_sub(x[i], x[j], p), p)
    end

    product = gf_div(y[i], product, p)
    t[0] = product

    for j in 0...d
      next if i == j
      (d - 1).downto(1) do |k|
        t[k] = gf_add(t[k], t[k - 1], p)
        t[k - 1] = gf_mul(t[k - 1], gf_sub(0, x[j], p), p)
      end
    end

    for j in 0...d
      coeff[j] = gf_add(coeff[j], t[j], p)
    end
  end

  coeff
end

def main
  if ARGV.length != 2
    puts "Error: Please provide a prime number and a file path."
  else
    p = ARGV[0].to_i
    path = ARGV[1]
    
    unless File.exist?(path)
      puts "Error: The file at '#{path}' doesn't exist!"
      return
    end

    begin
      d = 0
      x = []
      y = []

      File.open(path, 'r') do |file|
        file.each_line do |line|
          split = line.split(" ")
          raise ValueError if split.length != 2
          x << split[0].to_i
          y << split[1].to_i
          d += 1
        end
      end

      coeff = interpolate(d, x, y, p)
      puts coeff.join(" ")
    rescue
      puts "Error: Invalid input format"
    end
  end
end

main
